src/testing/slim_check/testable.lean

```diff
@@ -0,0 +1,490 @@
+
+/-- Test proposition `p` by randomly selecting one of the provided
+testable instance -/
+def combine_testable (p : Prop)
+  (t : list $ testable p) (h : 0 < t.length)
+: testable p :=
+⟨ λ min, have 0 < length (map (λ t, @testable.run _ t min) t),
+    by { rw [length_map], apply h },
+  gen.one_of (list.map (λ t, @testable.run _ t min) t) this ⟩
+
+/-- Once a property fails to hold on an example, look for smaller counter-examples
+to show the user -/
+def minimize [∀ x, testable (β x)] (x : α) (r : test_result (β x)) : lazy_list α → gen (Σ x, test_result (β x))
+| lazy_list.nil := pure ⟨x,r⟩
+| (lazy_list.cons x xs) := do
+  ⟨r⟩ ← uliftable.up $ testable.run (β x) tt,
+     if is_failure r
+       then pure ⟨x, convert_counter_example id r (psum.inl ())⟩
+       else minimize $ xs ()
+
+@[priority 5000]
+instance exists_testable {p : Prop}
+  (var var' : option string)
+  [testable (named_binder var (∀ x, named_binder var' $ β x → p))] :
+  testable (named_binder var' (named_binder var (∃ x, β x) → p)) :=
+⟨ λ min, do
+    x ← testable.run (named_binder var (∀ x, named_binder var' $ β x → p)) min,
+    pure $ convert_counter_example' exists_imp_distrib.symm x ⟩
+
+/-- Test a universal property by choosing sampleable examples to instantiate the
+bound variable with -/
+instance var_testable [has_to_string α] [sampleable α] [∀ x, testable (β x)]
+  (var : option string)
+: testable (named_binder var $ Π x : α, β x) :=
+⟨ λ min, do
+   uliftable.adapt_down (sample α) $
+   λ x, do
+     r ← testable.run (β x) ff,
+     uliftable.adapt_down (if is_failure r ∧ min then minimize _ _ x r (shrink x) else pure ⟨x,r⟩) $
+     λ ⟨x,r⟩, return $ match var with
+                      | none := add_to_counter_example (to_string x) ($ x) r
+                      | (some v) := add_var_to_counter_example v x ($ x) r
+                      end⟩
+
+@[priority 3000]
+instance unused_var_testable {β} [inhabited α] [testable β]
+  (var : option string)
+: testable (named_binder var $ Π x : α, β) :=
+⟨ λ min, do
+  r ← testable.run β min,
+  pure $ convert_counter_example ($ default _) r (psum.inr $ λ x _, x) ⟩
+
+@[priority 2000]
+instance subtype_var_testable {p : α → Prop} [has_to_string α] [sampleable (subtype p)]
+  [∀ x, testable (β x)]
+  (var var' : option string)
+: testable (named_binder var $ Π x : α, named_binder var' $ p x → β x) :=
+⟨ λ min,
+   do r ← @testable.run (∀ x : subtype p, β x.val) (slim_check.var_testable _ _ var) min,
+      pure $ convert_counter_example'
+        ⟨λ (h : ∀ x : subtype p, β x) x h', h ⟨x,h'⟩,
+         λ h ⟨x,h'⟩, h x h'⟩
+        r ⟩
+
+@[priority 100]
+instance decidable_testable {p : Prop} [decidable p] : testable p :=
+⟨ λ min, return $ if h : p then success (psum.inr h) else failure h [] ⟩
+
+section io
+
+variable (p : Prop)
+
+open nat
+
+variable {p}
+
+/-- execute `cmd` and repeat every time the result is `gave_up` or at most
+`n` times -/
+def retry (cmd : rand (test_result p)) : ℕ → rand (test_result p)
```

Comment:
If this is named `rand.retry` it will be nice for dot notation.

Reply:
That's true but it is fairly specific to `test_result` and if someone wants a general `retry` in `rand`, then the name is already taken.